### ISSUE TITLE
Cleaning up Basic and subclasses for evaluate/doit

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#17280]
TL;DR I think that SymPy's model for Basic is messy in places and I think that there could be significant benefits in cleaning it up. Below are some ideas.

SymPy's Basic evaluation model broadly works as follows:

Basic subclasses define `__new__` which
1. Usually sympifies the arguments
2. Possibly raises if the arguments are of the wrong type/value
3. Possibly evaluates by simplifying the arguments (e.g. `re(x+I) -> re(x)`)
4. Possibly evaluates to return a different object (`Mul(2, 3) -> Integer(6)`)
5. Otherwise sends the arguments to `Basic.__new__` to create the object, set
args etc.

Evaluation is typically controlled in a few ways:
1. Some `__new__` functions don't evaluate when given `evaluate=False` giving en "unevaluated object".
2. Others don't evaluate by default and always return unevaluated.
3. Unevaluated objects can be evaluated by calling doit.
4. doit has kwargs like `deep` which can control whether doit is called recursively on args.

That all seems like the nice model but it isn't implemented consistently across the SymPy code base and presumably by downstream authors/users writing their own Basic subclasses.

## 1. Inconsistent support for evaluate kwarg

There are 351 `__new__` methods in SymPy:
```shell
$ git grep 'def __new__' | wc -l
     351
```
I (crudely) count 3 `__new__` methods that have an evaluate parameter and 124 that have `**kwargs`. That means that the majority of these classes canot accept an evaluate kwarg. Of those `__new__` methods that take `**kwargs` many do not respect any `evaluate` kwarg and will evaluate regardless.

Classes like Derivative/Integral/Subs that don't evaluate by default are inconsistent in accepting an evaluate kwarg. For example Derivative respects evaluate=True but Integral raises. There is also no programmatic way to tell whether a particular class accepts an evaluate kwarg or whether it would or would not evaluate by default.

I propose that all Basic types should accept an evaluate=True kwarg. As a kwarg it should default to None but True, False and None should be accepted.  In the case of None a class attribute (`cls.autodoit`?) should be checked. This allows Integral etc to disable evaluation by default at the class level. Subclasses should then not define `__new__` but something else like `_eval_new` that would be called by `Basic.__new__` (if evaluate=True - perhaps this should be `_eval_doit` - see below).

Simpification should be handled in `Basic.__new__` perhaps controlled by a
`cls.simpifyargs` class attribute for subclasses that specifically want to
accept unsimpified args. It might also be worthwhile to have an _eval_validate
method that can be used to validate the arguments (raising if necessary) when
evaluate=True. A sketch of how this could look:

```python
class Basic:

    sympifyargs = True
    autodoit = True

    def __new__(cls, *args, evaluate=None, **kwargs):
        if cls.simpifyargs:
            args = tuple(simpify(a) for a in args)

        cls._eval_validate(*args, **kwargs) # Possibly raises

        # Current code of Basic goes here
        obj = object.__new__(cls)
        obj._assumptions = cls.default_assumptions
        obj._mhash = None  # will be set by __hash__ method.
        obj._args = args  # all items in args must be Basic objects

        # Now evaluate to something else
        if evaluate is None:
            evaluate = cls.autodoit
        if evaluate:
            obj = obj._eval_new()

        return obj


class Tuple(Basic):

    sympifyargs = False

    @classmethod
    def _eval_validate(cls, t):
        if not isinstance(t, tuple):
            raise TypeError("A tuple is expected")


class ImageSet(Set):

    autodoit = False

    @classmethod
    def _eval_validate(cls, func, sym, set):
        if not isinstance(func, Function):
            raise TypeError
        elif not isinstance(sym, Symbol):
            raise TypeError
        elif not isinstance(set, Set):
            raise TypeError

    def _eval_new(self):
        if self.func.is_Identity:
            return self.set
        return self
```

## 2. Patchy implementations of doit.

There are 76 doit methods in SymPy. None have an explicit deep kwarg. Of the 62 that do take kwargs many do not check for any deep flag so it isn't generally possible to request a deep/shallow doit. Some doit methods are always deep and others are always shallow. Some objects have doit methods that don't undo the effect of evaluate=False so that `func(*args, evaluate=False).doit() != func(*args)`.

This ImageSet for example doesn't evaluate because it hasn't asked the Integral to doit:
```julia
In [1]: ImageSet(Lambda(x, x**2), Interval(0, 1)).doit()
Out[1]: [0, 1]

In [2]: ImageSet(Lambda(x, Integral(x, x)), Interval(0, 1)).doit(deep=True)
Out[2]:
⎧⌠                  ⎫
⎨⎮ x dx | x ∊ [0, 1]⎬
⎩⌡                  ⎭
```
I propose that doit should be defined as a method on Basic that handles options like deep. Subclasses should not define doit and should instead define `_eval_doit` that evaluates only the current object. If this was done consistently then it would be possible to do things like `expr.doit(Subs)`, `expr.doit(subexpr)`, or `expr.doit(exclude=Integral)`.

A sketch for doit:
```python
class Basic:
    def doit(self, **kwargs):
        obj = self
        if kwargs.get('deep', True):
            args = (a.doit(**kwargs) for a in obj.args)
            if args != obj.args:
                obj = type(obj)(*args, evaluate=False)
        exclude = kwargs.get('exclude', ())
        if not any(isinstance(obj, cls) for cls in exclude):
            return obj._eval_doit()
```

## 3. Refine and new assumptions

Currently there is no way to incorporate the new assumptions in evaluation except using global state as in `with assuming(assumptions)` which is not good as a general solution because globals are bad, not threadsafe etc. The code for refine is a parallel implementation of the code for normal evaluation. This means that refine cannot give results consistent with the old assumptions without reimplementing all of the evaluation code (`Mul.flatten` etc).

If evaluation was controlled in a more systematic way (as described above) then it would be possible to add assumptions arguments in key places in order to accommodate for the new assumptions in this setup. Ideally the code to evaluate an object would be the same as the code to refine the object except that additional assumptions get passed to ask internally. In particular `_eval_new` could have an assumptions argument and an `_eval_refine` method for Basic could be something like
```python
    def _eval_refine(self, assumptions):
        args = tuple(refine(a, assumptions) for a in self.args)
        return self._eval_new(*args, assumptions=assumptions)
```
The `_eval_new` can use `ask` internally instead of the old assumptions.

(This last part about refine is something I am less sure about. In particular I wonder about the fact that the new assumptions are built on top of Basic so I'm not sure how this would bootstrap. Maybe these changes belong at Expr rather than Basic but refine could be equally useful for non-Expr classes like sets.)

## 4. Backwards compatibility

I don't think it's reasonable to expect every author of a `doit/__new__` method to get all of the details right so it's important that a simpler subclassing interface is provided that naturally gets the details right.  Ideally I want to get to a state where implementing `doit/__new__` directly is discouraged (just like it is for Function subclasses which should define eval instead). However implementing `__new__` will still need to be supported for backwards compatiblity. It's also important that this work can be done incrementally rather than as a mega-patch.

I propose that the changes described above should be somehow opt-in and backwards compatible. This means in particular that anyone who previously wrote a class that defines `__new__` needs `Basic.__new__` to not trigger any evaluation. Maybe there should be some kind of version flag like `cls.basic_v2` or something to know whether `Basic.__new__` should be a no op. Perhaps testing for the presence of `_eval_new` would be sufficient for this.

If this was done in the right way I think it would be possible to make these changes incrementally across the SymPy codebase leading to a cleaner overall structure.

**Comment by smichr** [comment]
> 2\. Possibly raises if the arguments are of the wrong type/value

See also discussion [here](https://groups.google.com/forum/#!topic/sympy/2-exgd0_5Io)

**Comment by oscarbenjamin** [comment]
I was hoping for more discussion on this. Should I interpret this as "open a PR and we'll talk then"?

**Comment by jksuom** [comment]
> open a PR and we'll talk then

I guess that might be the best way to proceed. There are several items, and probably each one would need at least one PR.

**Comment by oscargus** [comment]
I'll have to read it a bit more carefully, but isn't a problem that even with the new model, people can still ignore it and implement `__new__` methods that doesn't do as expected?

I think my real questions is: why do we need a new method instead of just fix all the current classes? (We will still need to fix *ALL* classes and adapt them to the new method, keeping the old limits it to the broken ones.)

(But I may have to read it more carefully some other time as the answer may be there.)

**Comment by oscarbenjamin** [comment]
> why do we need a new method instead of just fix all the current classes?

The current design apparently makes it very easy to get things wrong. I would like a new design that makes it easier to get things right. Moving evaluate=True/False out of the hands of subclass authors should mean that it automatically works in most cases. The same goes for sympification etc. Likewise most doit methods either have to duplicate handling e.g. deep=True/False or just don't bother.

**Comment by sylee957** [comment]
I've also found that there is an another key `check` used in matrix expressions like in #16839, because of some need to return some mathematically nonsensical object in mid-computation.

If `check` key is added to the evaluation control hierarchy

1. Sympify the args
2. Check the arg types (if `check=True`)
3. Do a computationally cheap canonicalization (if `evaluate=True`)
4. Do a computationally expensive evaluation (if `doit=True`)

And if step 1 is skipped, the resulting object may not work with sympy traversal methods or printing functions.
If step 2 is skipped, the resulting object should not have assumptions valid

If there are other evaluation controls for other modules, it should be extended further.

**Comment by oscarbenjamin** [comment]
I would prefer to have a way to avoid creating mathematically nonsensical objects. The issue in #16839 is about substituting matrices for non-matrices. Maybe that should be done with xreplace or something else rather than subs since it isn't a mathematically valid substitution.

**Comment by asmeurer** [comment]
I agree with the ideas here. It's always bothered me that sympify wasn't called automatically by Basic. This would also be a good way to get instantantiation sympification to use _sympify instead of sympify.

**Comment by mcpl-sympy** [comment]
Being attempted in #18873 as well.

**Comment by oscarbenjamin** [comment]
For pickling we need a function that can create a Basic object from valid args without any evaluation. Ideally that would be `cls.__new__` but there are too many `__new__` methods with automatic evaluation to use that. We can't use `evaluate=False` because it's not consistently supported in the codebase and presumably also in downstream code. It also can't be `Basic.__new__(cls, ...)` because too many `Basic` subclasses perform other initialisation after calling `Basic.__new__` e.g. by setting attributes on the object e.g.:
```python
class Thing(Basic):
    def __new__(cls, arg1, arg2):
        obj = super().__new__(cls, arg1, arg2)
        obj.thing = func(arg1)
        return obj
```

We should have a new class method `Basic.new` that creates a new instance with those args and never evaluates e.g.:
```python
class Basic:
    def __new__(cls, *args):
        obj = super().__new__(cls)
        obj._args = args
        return obj

    @property
    def func(self):
        return type(self)

    @property
    def args(self):
        return self._args

    # We pass args and kwargs as a pair of positional arguments so that this
    # can be used with the __reduce_ex__ pickling method that only supports
    # positional arguments.
    @classmethod
    def new(cls, args, kwargs=None):
        # Subclasses opt in by implementing _init
        _init = getattr(cls, '_init', None)

        # Backwards compat fallback
        if _init is None:
            if kwargs:
                return cls(*args, **kwargs)
            else:
                return cls(*args)

        # Don't use super and don't sympify.
        # If the args are not Basic or if kwargs are needed then a subclass
        # will need to override this method. We need to support kwargs for
        # e.g. Symbol but otherwise using kwargs with Basic is discouraged.
        obj = Basic.__new__(cls, *args)
        obj._init(*args)
        return obj

    # For pickling we use self.new rather than self.func
    def __reduce_ex__(self, protocol):
        return self.new, self.__getnewargs_ex__()

    # Some subclasses might use different args for self.func compared to
    # self.new e.g. atoms like Integer, Rational, etc. Also some cases like
    # Symbol need to support keyword arguments. In those cases this method
    # will need to be overridden.
    def __getnewargs_ex__(self):
        return self.args, {}


class Thing(Basic):
    def __new__(cls, *args, *, evaluate=False):
        args = tuple(map(_sympify, args))

        if evaluate:
            if any(arg == 0 for arg in args):
                return OtherThing()

        # Use cls.new here rather than super().__new__ and pass args rather
        # than *args.
        return cls.new(args)

    def _init(obj, arg1, arg2):
        obj.thing = func(arg1)


class Atom(Basic):
    @classmethod
    def new(cls, args, kwargs):
        # Don't pass args to Basic.__new__ for Atom:
        obj = Basic.__new__(cls)
        obj._init(*args, **kwargs)
        return obj


class Symbol(Basic):
    def __new__(cls, *args, **assumptions):
        return cls.new(args, assumptions)

    def _init(obj, name, **asumptions):
        obj._name = name
        obj._assumptions = assumptions

    def __getnewargs_ex__(self):
        return (self.name,), self._assumptions
```
It might be easier to implement something like `Basic.new` as a transition towards the ideas in the OP.
